drivers/interrupt_controller/intc_scr_ipic.c

```diff
@@ -0,0 +1,209 @@
+/**
+ * Copyright (c) 2023 Syntacore
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * @file intc_scr_ipic.c
+ * @author psk-sc
+ * @brief Syntacore Integrated Programmable Interrupt Controller (IPIC) Interface
+ *        for RISC-V processors
+*/
+
+#define DT_DRV_COMPAT riscv_scr_ipic
+
+/******************************************************************************
+ *                              Includes section
+ */
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/init.h>
+#include <zephyr/arch/riscv/csr.h>
+#include <zephyr/device.h>
+
+#include <zephyr/sw_isr_table.h>
+#include <zephyr/drivers/interrupt_controller/riscv_ipic.h>
+#include <zephyr/irq.h>
+
+/******************************************************************************
+ *                              Defines section
+ */
+#define PLF_IPIC_STATIC_LINE_MAPPING CONFIG_IPIC_STATIC_LINE_MAPPING
+#define PLF_IPIC_IRQ_LN_NUM (CONFIG_EXT_IPIC_IRQ_LN_NUM)
+#define PLF_IPIC_IRQ_VEC_NUM PLF_IPIC_IRQ_LN_NUM
+#define IPIC_IRQ_LN_VOID  PLF_IPIC_IRQ_LN_NUM
+
+#define IPIC_IRQ_PENDING       (1 << 0)
+#define IPIC_IRQ_ENABLE        (1 << 1)
+#define IPIC_IRQ_LEVEL         (0 << 2)
+#define IPIC_IRQ_EDGE          (1 << 2)
+#define IPIC_IRQ_INV           (1 << 3)
+#define IPIC_IRQ_MODE_MASK     (3 << 2)
+#define IPIC_IRQ_CLEAR_PENDING IPIC_IRQ_PENDING
+
+#define IPIC_IRQ_IN_SERVICE (1 << 4) // RO
+#define IPIC_IRQ_PRIV_MASK  (3 << 8)
+#define IPIC_IRQ_PRIV_MMODE (3 << 8)
+#define IPIC_IRQ_PRIV_SMODE (1 << 8)
+#define IPIC_IRQ_LN_OFFS    (12)
+
+#ifndef PLF_IPIC_MBASE
+#define PLF_IPIC_MBASE (0xbf0)
+#endif
+#define IPIC_CISV  (PLF_IPIC_MBASE + 0)
+#define IPIC_CICSR (PLF_IPIC_MBASE + 1)
+#define IPIC_IPR   (PLF_IPIC_MBASE + 2)
+#define IPIC_ISVR  (PLF_IPIC_MBASE + 3)
+#define IPIC_EOI   (PLF_IPIC_MBASE + 4)
+#define IPIC_SOI   (PLF_IPIC_MBASE + 5)
+#define IPIC_IDX   (PLF_IPIC_MBASE + 6)
+#define IPIC_ICSR  (PLF_IPIC_MBASE + 7)
+#define IPIC_IER   (PLF_IPIC_MBASE + 8)
+#define IPIC_IMAP  (PLF_IPIC_MBASE + 9)
+#define IPIC_VOID_VEC 16
+
+#define MK_IRQ_CFG(line, mode, flags) ((mode) | (flags) | ((line) << IPIC_IRQ_LN_OFFS))
+
+/******************************************************************************
+ *                              Interface implementation section
+ */
+static void ipic_irq_enable(unsigned irq_vec);
+
+static void ipic_irq_disable(unsigned irq_vec);
+
+static unsigned long ipic_soi(void);
+
+static void ipic_eoi(void);
+
+static void ipic_irq_reset(int irq_vec);
+
+static int ipic_irq_setup(int irq_vec, int line, int mode, int flags);
+
+static unsigned long ipic_irq_current_vector(void);
```

Comment:
why?

Reply:
again, why?

Reply:
First of all, thanks for all of your commentary
This was done to be able to put major functions before utility functions. If it's not appropriate, I cant put utility functions on the first place, and remove unneeded declarations. I can't really see another options, so, please, comment, if you have some